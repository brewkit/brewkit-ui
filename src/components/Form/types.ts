import { ReactNode, ReactElement } from 'react';


export interface Props {
    children?: ReactNode,
}


export interface FormComponent {
    Group?: any,
    Label?: any,
    Field?: any,
    HelperText?: any,
    (props: Props): ReactElement,
}
