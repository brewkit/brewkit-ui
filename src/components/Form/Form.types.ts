import { FormHTMLAttributes } from 'react';
import { SubmitHandler, UseFormReturn } from 'react-hook-form';


export type FormProps<T = never> = FormHTMLAttributes<HTMLFormElement> & {

    /**
     * function to be called on form submission
     */
    onSubmit: SubmitHandler<T>,

    /**
     * configuration options to be passed to the underlying RHF `useForm` hook
     * (https://react-hook-form.com/api#useForm)
     */
    useFormMethods?: UseFormReturn,


};
