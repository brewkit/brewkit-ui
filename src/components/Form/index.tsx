import * as React from 'react';
import { useForm, FormProvider } from 'react-hook-form';
import { AnimateSharedLayout } from 'framer-motion';
import { Props } from './types';


/**
 * `Form` wraps around underlying `Form` based components and provides a context provider for them to bind to. It also
 * handles the form's submission and can use validation schema to validate the content.
 *
 * > This component is not found in Material UI.
 */
export const Form = React.forwardRef(({
    children,
    onSubmit,
    ...otherProps
}: Props, ref: React.Ref<any>): React.ReactElement => {


    const methods = useForm({
        mode: 'onChange',
    });


    return (
        <FormProvider {...methods}>
            <form noValidate onSubmit={methods.handleSubmit(onSubmit)} ref={ref} {...otherProps}>
                <AnimateSharedLayout>
                    {children}
                </AnimateSharedLayout>
            </form>
        </FormProvider>
    );


});


Form.displayName = 'Form';


export default Form;
