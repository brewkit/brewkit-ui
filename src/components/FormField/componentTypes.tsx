import Checkbox from '../Checkbox';
import Radio from '../Radio';
import TextField from '../TextField';
import Switch from '../Switch';

/* Map our input components for easier access */
export default {
    checkbox: Checkbox,
    radio: Radio,
    switch: Switch,
    text: TextField,
    password: TextField,
    number: TextField,
};
