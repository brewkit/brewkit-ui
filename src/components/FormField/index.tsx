import * as React from 'react';
import { useFormContext } from 'react-hook-form';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { AnimatePresence, motion } from 'framer-motion';
import Checkbox from '@components/Checkbox';
import Radio from '@components/Radio';
import TextField from '@components/TextField';
import Switch from '@components/Switch';
import { Props } from './types';


/* Map our input components for easier access */
const components = {
    checkbox: Checkbox,
    radio: Radio,
    switch: Switch,
    text: TextField,
    password: TextField,
    number: TextField,
};


/**
 * `FormField` is an abstraction of most inputs, additionally binding the rendered input to a `Form` so they can be
 * tracked and validated.
 *
 * > This component is not found in Material UI.
 */
export const FormField = React.forwardRef(({
    type = 'text',
    label,
    name,
    validation = {},
    helperText,
    ...otherProps
}: Props, ref: React.Ref<any>): React.ReactElement => {


    const { register, errors } = useFormContext();
    const Component: any = components[type] || TextField;


    /**
     * configure our Framer animation
     */
    const fadeAnim = {
        initial: { opacity: 0, y: -5 },
        animate: { opacity: 1, y: -0 },
        exit: { opacity: 0 },
    };


    /**
     * determine what to render in the helper text. Keys have to be changed based on content ot ensure the component
     * updates correctly.
     */
    function getHelperText(): React.ReactNode {

        let content: React.ReactNode = ' ';
        let key = 1;

        /** if there is an error, return that */
        if (errors[name]?.message) {
            content = errors[name]?.message;
            key = 2;
        }

        /** else, return helperText if provided */
        if (helperText) {
            content = helperText;
            key = 3;
        }

        if (key === 1) return <AnimatePresence />;

        return (
            <AnimatePresence>
                <motion.span key={key} layout style={{ display: 'inline-block' }} {...fadeAnim}>
                    {content}
                </motion.span>
            </AnimatePresence>
        );

    }


    /**
     * if the component is a TextField, we just use that.
     */
    if (Component === TextField) return (
        <Component
            error={Boolean(errors[name])}
            helperText={getHelperText()}
            inputRef={register(validation)}
            label={label}
            name={name}
            type={type}
            {...otherProps}
        />
    );


    /**
     * otherwise, we wrap it in a FormControlLabel
     */
    return (
        <FormControlLabel
            control={<Component name={name} {...otherProps} />}
            inputRef={register(validation)}
            label={label}
        />
    );


});


FormField.displayName = 'FormField';


export default FormField;
