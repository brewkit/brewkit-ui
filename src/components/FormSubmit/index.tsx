import * as React from 'react';
import { useFormContext } from 'react-hook-form';
import Button from '@components/Button';
import { Props } from './types';


/**
 * `FormSubmit` is a submit button with additional bindings to a parent `Form` to allow for better control and
 * validation. It automatically disables itself while it's parent `Form` is incomplete.
 */
export const FormSubmit = React.forwardRef(({
    children = 'Submit',
    ...otherProps
}: Props, ref: React.Ref<any>): React.ReactElement => {


    const { formState } = useFormContext();


    return (
        <Button
            disabled={!(formState.isValid)}
            ref={ref}
            type="submit"
            {...otherProps}
        >
            {children}
        </Button>
    );


});


FormSubmit.displayName = 'FormSubmit';


export default FormSubmit;
