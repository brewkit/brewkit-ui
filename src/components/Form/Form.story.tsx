import * as React from 'react';
import { useForm } from 'react-hook-form';
import Form from '@components/Form';
import FormField from '@components/FormField';
import FormSubmit from '@components/FormSubmit';


export default {
    component: Form,
    parameters: {
        // componentSubtitle: <Flag color="success">Stable</Flag>,
    },
    title: 'Forms/Form',
};


export const General = (): React.ReactElement => {

    const validation = {
        required: {
            value: true,
            message: 'This field is required.',
        },
        minLength: {
            value: 10,
            message: 'A length of 10 or more is required.',
        },
    };


    const handleSubmit = (): Promise<string> => new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve('foo');
        }, 2000);
    });


    return (
        <Form onSubmit={handleSubmit} style={{ display: 'grid', gridGap: '1rem' }} >
            <FormField autoComplete="new" label="text1" name="text1" placeholder="text" type="text" validation={validation} />
            <FormField label="Switch" name="switch" type="switch" validation={validation} />
            <FormField label="Checkbox" name="checkbox" type="checkbox" validation={validation} />
            <FormSubmit>Submit</FormSubmit>
        </Form>
    );

};


export const CustomFormMethods = (): React.ReactElement => {

    const formMethods = useForm({ mode: 'onChange' });

    const validation = {
        required: {
            value: true,
            message: 'This field is required.',
        },
        minLength: {
            value: 10,
            message: 'A length of 10 or more is required.',
        },
    };


    const handleSubmit = (values: any): Promise<any> => new Promise((resolve, reject) => {
        setTimeout(() => {
            console.log(values);
            resolve('foo');
        }, 2000);
    });


    return (
        <Form onSubmit={handleSubmit} style={{ display: 'grid', gridGap: '1rem' }} useFormMethods={formMethods}>
            <FormField autoComplete="new" label="text1" name="text1" placeholder="text" type="text" validation={validation} />
            <FormField label="Switch" name="switch" type="switch" validation={validation} />
            <FormField label="Checkbox" name="checkbox" type="checkbox" validation={validation} />
            <FormSubmit>Submit</FormSubmit>
        </Form>
    );

};
