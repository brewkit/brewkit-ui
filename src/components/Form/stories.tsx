import * as React from 'react';
import { text, boolean } from '@storybook/addon-knobs';
import Form from '@components/Form';
import FormField from '@components/FormField';
import FormSubmit from '@components/FormSubmit';


export default {
    component: Form,
    parameters: {
        // componentSubtitle: <Flag color="success">Stable</Flag>,
    },
    title: 'Forms/Form',
};


export const General = (): React.ReactElement => {

    const validation = {
        required: {
            value: true,
            message: 'This field is required.',
        },
        minLength: {
            value: 10,
            message: 'A length of 10 or more is required.'
        }
    };

    return (
        <Form style={{ display: 'grid', gridGap: '1rem' }} onSubmit={(values: any) => console.log(values)} >
            <FormField label="text1" name="text1" type="text" placeholder="text" autoComplete="new" validation={validation} />
            <FormField label="Switch" name="switch" type="switch" validation={validation} />
            <FormField label="Checkbox" name="checkbox" type="checkbox" validation={validation} />
            <FormSubmit>Submit</FormSubmit>
        </Form>
    );

}
