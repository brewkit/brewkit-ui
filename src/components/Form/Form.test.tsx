// @ts-nocheck
import * as React from 'react';
import { create } from 'react-test-renderer';
import Form from '.';
import FormField from '../FormField';
import FormSubmit from '../FormSubmit';

it('Renders Form component', () => {
    const comp = create(
        <Form>
            <FormField label="text1" name="text1" type="text" placeholder="text" />
            <FormField label="Switch" name="switch" type="switch" />
            <FormField label="Checkbox" name="checkbox" type="checkbox" />
            <FormSubmit>Submit</FormSubmit>
        </Form>
    ).toJSON();

    expect(comp).toMatchSnapshot();
});
