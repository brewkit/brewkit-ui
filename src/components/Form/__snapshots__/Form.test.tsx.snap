// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Form component 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="MuiFormControl-root MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink={false}
    >
      text1
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        disabled={false}
        name="text1"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="text"
        required={false}
        type="text"
      />
    </div>
  </div>
  <label
    className="MuiFormControlLabel-root"
  >
    <span
      className="MuiSwitch-root"
    >
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-6 MuiSwitch-switchBase MuiSwitch-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            className="PrivateSwitchBase-input-9 MuiSwitch-input"
            name="switch"
            onChange={[Function]}
            type="checkbox"
          />
          <span
            className="MuiSwitch-thumb"
          />
        </span>
      </span>
      <span
        className="MuiSwitch-track"
      />
    </span>
    <span
      className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    >
      Switch
    </span>
  </label>
  <label
    className="MuiFormControlLabel-root"
  >
    <span
      aria-disabled={false}
      className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-6 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <span
        className="MuiIconButton-label"
      >
        <input
          className="PrivateSwitchBase-input-9"
          data-indeterminate={false}
          name="checkbox"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="Checkbox-controlBase-10 Checkbox-uncheckedControl-11"
        >
          <div
            style={
              Object {
                "opacity": 0,
              }
            }
          />
        </div>
      </span>
    </span>
    <span
      className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    >
      Checkbox
    </span>
  </label>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text Button-root-13"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="MuiButton-label"
    >
      <div
        className="Button-content-14"
      >
        <span>
          Submit
        </span>
      </div>
    </span>
  </button>
</form>
`;
